src/Chapter4.hs

```diff
@@ -629,7 +670,22 @@ Can you implement a monad version of AND, polymorphic over any monad?
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+andM a b = fmap (&&) a <*> b -- only needs applicative
+
+andM2 :: (Monad m) => m Bool -> m Bool -> m Bool
+andM2 x y=(x >>= (\a -> pure (&& a)))<*>y -- follows hint but quite ugly
+
+andM3 :: (Monad m) => m Bool -> m Bool -> m Bool
+andM3 = liftA2 (&&) -- kind of cheating...
```

Comment:
The ordinary `&&` if lazy on the second argument. Meaning that if it meets `False` as the first argument, there is no need to calculate another arg. In `andM` we would like to keep the same behaviour, so the

```
Just False `andM` _
```

should be `Just False` :slightly_smiling_face: 


If you don't use the `>>=` operator, then the `Monad` constraint is too strong, and it is enough to specify only `Applicative`. However, with `Applicative` you are not able to implement a lazier version of `andM`.

Reply:
Phew, this really drove me crazy! The real final boss as far as I'm concerned
Eventually I got there and I'm quite pleased by the solution
Thanks a lot for the comment above, very insightful!